### ISSUE TITLE
Build failure: feishin

### ISSUE THREAD

**Issue by lunik1** [NixOS/nixpkgs#262739]
feishin fails to build as it depends on `electron_24`, which is now EOL

### Steps To Reproduce

1. build feishin

### Notify maintainers

@onny 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.7, NixOS, 23.05 (Stoat), 23.05.20231015.0e1cff5`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.17.0`
 - channels(root): `"nixos"`
 - channels(lunik1): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by Animeshz** [comment]
Also, logseq (due to electron_24).

**Comment by avh4** [comment]
FYI, feishin-0.5.1 has been released upstream, which uses Electron 25.  (though I tried updating it in my local nixpkgs and just got a blank white window, so I'm not sure if something else is wrong or just if my local environment is weird in some way w/r to openGL drivers)

**Comment by minijackson** [comment]
@avh4 I can confirm your issue, I also have a white screen with Feishin 0.5.1 and Electron 25.

**Comment by MikaelFangel** [comment]
Reference #263231

**Comment by onny** [comment]
Feishin 0.5.2 with support for Electron v27 but issue still persists. This is propably an issue on NixOS only?

**Comment by lunik1** [comment]
Yeah, the version of electron used needs to be bumped in https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/audio/feishin/linux.nix (`electron_24` → `electron_27`). However, when I do this and bump the version feishin does not work correctly, only showing a blank windown when I launch it.

**Comment by lunik1** [comment]
The AppImage of 0.5.3 seems to work, I've made a package in [my NUR](https://github.com/lunik1/nur-packages/blob/master/pkgs/feishin-appimage/default.nix). `x86_64-linux` only.

**Comment by eclairevoyant** [comment]
For the blank screen, try the following, in order
* Delete the ~/.config/(whatever)/GPUCache folder, then running the app
* Run the app with `--disable-gpu`
* Unset `NIXOS_OZONE_WL` before running the app

**Comment by lunik1** [comment]
No luck with any of those solutions for me

**Comment by jlbribeiro** [comment]
> No luck with any of those solutions for me

@lunik1 You're right, just bumping Electron wasn't enough.
(Forgive me for using this issue for documenting the rationale behind #303638.)

This in particular was due to
- Feishin choosing the [resources path based on `app.isPackaged`](https://github.com/jeffvli/feishin/blob/3976f5e5bfd4719a8882aa2601f41c99392def8e/src/main/main.ts#L103-L105), combined with
- [`process.resourcesPath` currently not being configurable](https://github.com/electron/electron/issues/31121) which, in our case, means it always points to an `electron`-global resources path (since we're re-using the same `electron` binary for packaging multiple applications).

[`app.isPackaged` depends on whether the executable is named `electron`](https://github.com/electron/electron/issues/35153#issuecomment-1202718531); if `executable ~= electron`, it assumes the app is **not** packaged; in our case that means `app.isPackaged` is always false, causing it to fail due to not finding assets, such as the app icon.

After fixing those, it failed because
- the final packaged app depends on additional dependencies specified in [a separate `release/app/package.json`](https://github.com/jeffvli/feishin/blob/3976f5e5bfd4719a8882aa2601f41c99392def8e/release/app/package.json); in the original packaging scripts, this gets installed by `electron-builder install-app-deps` _magic_. In this particular case, it would fail with a missing `mpris-service`, as mentioned in the [initial draft PR](https://github.com/NixOS/nixpkgs/pull/288210#issue-2129430059).

#303638 should have fixed this by
- forcing the app to be recognized as being packaged by [setting `ELECTRON_FORCE_IS_PACKAGED=1`](https://github.com/NixOS/nixpkgs/blob/623ac957cb99a5647c9cf127ed6b5b9edfbba087/pkgs/applications/audio/feishin/default.nix#L92);
- [replacing `process.resourcesPath` with an app-specific resources path](https://github.com/NixOS/nixpkgs/blob/623ac957cb99a5647c9cf127ed6b5b9edfbba087/pkgs/applications/audio/feishin/default.nix#L42-L44); and
- [merging the release app dependencies with the final `node_modules`](https://github.com/NixOS/nixpkgs/blob/623ac957cb99a5647c9cf127ed6b5b9edfbba087/pkgs/applications/audio/feishin/default.nix#L47-L68).

Please let me know if it didn't fix it for you.
